### PR TITLE
Maximum length for summaries for Guilds and Challenges (#9244)

### LINKED ISSUE #9244: API must enforce maximum length for summaries for Guilds and Challenges
Guilds and Challenges have a summary field and the website enforces a maximum length for it:
https://github.com/HabitRPG/habitica/blob/6838b7d0a60c1b5b089ae5dab17f9cf10e09b937/website/common/script/constants.js#L9-L10

The API's routes for creating and editing Guilds and Challenges should enforce the same maximum.

Note that currently the API does not make the summary field mandatory - that is desired behaviour at the moment because a mandatory summary would be a breaking change for the API. Because of this, the API should enforce the maximum length only if a summary is provided. Don't make the summary mandatory.

There should also be tests that ensure that IF a summary is provided, it is not too long.

### PR DESCRIPTION
**Pull request by marvinrabe**
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #9244

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)

Added express validations for summary field with tests.

[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: b3eeebf4-b03e-48dd-9e3c-abb27410d1d4

### CODE REVIEW

**Comment by marvinrabe** [comment]
Travis failed because of:

>    1) GET /challenges/:challengeId/members
>        supports using req.query.lastId to get more members:
>      Error: Timeout of 8000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/travis/build/HabitRPG/habitica/test/api/v3/integration/challenges/GET-challenges_challengeId_members.test.js)

Seems like a random error. Not my fault. 😔

File: test/api/v3/integration/groups/PUT-groups.test.js [hunk 1]
```diff
@@ -3,6 +3,7 @@ import {
   generateUser,
   translate as t,
 } from '../../../../helpers/api-v3-integration.helper';
+import {MAX_SUMMARY_SIZE_FOR_GUILDS} from '../../../../../website/common/script/constants.js';
```
Comment (paglias) [inline test/api/v3/integration/groups/PUT-groups.test.js:6]: shouldn't this check be at the model level instead of doing it for every route?

Comment (marvinrabe) [reply test/api/v3/integration/groups/PUT-groups.test.js:6]: See: https://github.com/HabitRPG/habitica/issues/9244#issuecomment-351550431

> So, I think the best start here is to validate length during the api request.

Comment (marvinrabe) [reply test/api/v3/integration/groups/PUT-groups.test.js:6]: I dont know what the reason for this is. Though if requested I could move it to the model level instead. Seems more reasonable anyway…

Comment (paglias) [reply test/api/v3/integration/groups/PUT-groups.test.js:6]: Hmm I see, I think the problem might be that we don't want to start throwing errors if someone had a longer summery before the validation is introduced

**Comment by Alys** [comment]
@marvinrabe Just to confirm, this won't give an error if the summary does not exist, right?

**Comment by marvinrabe** [comment]
@alys No, it wont. There are tests for the "no summary case" and they are green. For example this one: https://github.com/HabitRPG/habitica/blob/bbbd1f9f73ab69905f9e5957c7055a6206a14d30/test/api/v3/integration/challenges/POST-challenges.test.js#L318

**Review by paglias — COMMENTED** [review]
The code is okay but honestly I'm worried about the validation not being in the model. @Alys what do we want to do with summaries that are too long?

If we could cut them with a migration then we could move the validation to the model

**Comment by Alys** [comment]
We can certainly fix any too-long summaries before this goes live but I don't think it's worth writing a migration script. There'll be so few that we can easily handle them manually and we'd want to adjust them manually anyway (e.g., reduce unneeded words rather than cut at after 250 characters). I've searched challenges and guilds just now and found only one that had a long summary, which I've shortened. I'll repeat the search before go-live and again just after (if someone could remind me when this is merged and again when deployed, that would be helpful but I'll make a note to check every day or so).
Sample code:
`db.groups.find({"summary": { "$regex": /^.{251,}$/} },{_id:1, summary:1},{multi:true});`

However if the validation is in the model, doesn't that mean it will be checked every time anything happens with the guild (e.g., a user viewing it), rather than just when it's needed (when the summary is changed)? If that's what you want, I'm fine with it but it seems like overkill. I could well be misunderstanding though!

**Comment by paglias** [comment]
The validation at the model level would only happen when the field is modified so it's not a problem.

@marvinrabe can you change the code to have the validation happen at the model level?

**Comment by marvinrabe** [comment]
@paglias Sure!

**Comment by beffymaroo** [comment]
@marvinrabe Are you still working on this issue? Let us know when you have a moment. Thank you!

**Pull request closed without merging**